src/viztracer/modules/vcompressor/vc_dump.c

```diff
@@ -38,6 +38,57 @@
     fputc('\0', fptr);                                                       \
 }
 
+#define READ_ENCODED_INT64(num, fptr)                                        \
+{                                                                            \
+    uint8_t flag = 0;                                                        \
+    uint8_t encoded_int_low = 0;                                             \
+    uint32_t encoded_int_32 = 0;                                             \
+    uint64_t encoded_int_64 = 0;                                             \
+    READ_DATA(&flag, uint8_t, fptr);                                         \
+    switch (flag & 0xC0)                                                     \
+    {                                                                        \
+        case TS_6_BIT:                                                       \
+            num = flag & 0x3F;                                               \
+            break;                                                           \
+        case TS_14_BIT:                                                      \
+            READ_DATA(&encoded_int_low, uint8_t, fptr);                      \
+            num = ((flag & 0x3F) << 8) | encoded_int_low;                    \
+            break;                                                           \
+        case TS_32_BIT:                                                      \
+            READ_DATA(&encoded_int_32, uint32_t, fptr);                      \
+            num = encoded_int_32;                                            \
+            break;                                                           \
+        case TS_64_BIT:                                                      \
+            READ_DATA(&encoded_int_64, uint64_t, fptr);                      \
+            num = encoded_int_64;                                            \
+            break;                                                           \
+        default:                                                             \
+            printf("shouldn't be here!!!");                                  \
+            break;                                                           \
+    }                                                                        \
+}                                                                            \
+
+#define WRITE_ENDOCED_INT64(dump_int, fptr)                                  \
+{                                                                            \
+    if((dump_int & 0x3F) == dump_int){                                       \
```

Comment:
You are using 5 bytes for 32 bit value and 9 bytes for 64 bit value now. Notice that all the values could fit in a 64-bit integer with ns as unit. You did `*100` instead of `*1000` so the unit is `10ns`, which is a reasonable choice for resolution(let's see if it causes timestamp conflict). That means, you have 3 bits (*8) redundancy. So, in the worst case, you can fit your value in 61 bits. That said, you are wasting a byte for 32/64 bit value.

What I would suggest here is to create a new MACRO `PEEK` (as compared to READ_DATA). Instead of read a byte in and change the fptr, it "peeks" the next byte by reading it in and set the fptr back (using `fseek(fp, -1, SEEK_CURR)`). This way, you can check the flag first, determine what's the size to read in, and read the data in with `READ_DATA`. Masking out the flag part would be simpler as well.

However, there would be a case, when the ts is negative and probably needs all 64 bits. I would suggest that we have a special flag - `0x00` to indicate that. In theory, `0x00` should be an invalid value because all the ts should be distinct and all the durations should be greater than 0.